src/gui/shell.cpp

```diff
@@ -1212,52 +1211,70 @@ void Shell::mouseMoveEvent(QMouseEvent *ev)
 	}
 }
 
+static int normalize(float x)
+{
+    if (x == 0.0f) return 0;
+    else if (x < 0) return -1;
+    else return 1;
+}
+
 void Shell::wheelEvent(QWheelEvent *ev)
 {
 	if (!m_attached || !m_mouseEnabled) {
 		return;
 	}
-#ifdef Q_OS_MAC
-	// For some reason <ScrollWheel*> scrolls multiple lines at once
-	// we have to account for it, to make sure that pixelDelta() is used correctly.
-	const int scroll_columns = 6;
-	const int scroll_rows = 3;
-	// Minimal scroll step
-	int scroll_step_x = cellSize().width() * scroll_columns;
-	int scroll_step_y = cellSize().height() * scroll_rows;
-	// Total scroll delta considering previous events
-	QPoint total_delta = m_mouse_wheel_delta_fraction + ev->pixelDelta();
-	// Delta rounded to a minimal scroll step
-	QPoint cell_delta(total_delta.x() / scroll_step_x, total_delta.y() / scroll_step_y);
-	// Save remainder for future events
-	m_mouse_wheel_delta_fraction = total_delta - QPoint(cell_delta.x() * scroll_step_x, cell_delta.y() * scroll_step_y);
-
-	int horiz = cell_delta.x();
-	int vert = cell_delta.y();
+#if (QT_VERSION < QT_VERSION_CHECK(5, 5, 0))
+	int defaultDeltasPerStep = 120;
 #else
-	int horiz, vert;
-	horiz = ev->angleDelta().x();
-	vert = ev->angleDelta().y();
+	int defaultDeltasPerStep = QWheelEvent::DefaultDeltasPerStep;
+#endif
+
+	QPointF scroll_delta_f(float(ev->angleDelta().x()) / defaultDeltasPerStep,
+						   float(ev->angleDelta().y()) / defaultDeltasPerStep);
+
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 7, 0))
+	if (ev->inverted()) {
+		scroll_delta_f = -scroll_delta_f;
+	}
 #endif
 
-	if (horiz == 0 && vert == 0) {
+	// Reset scroll remainder if we change scroll direction;
+	int direction_x = normalize(scroll_delta_f.x());
+	if (direction_x != m_scroll_last_direction.x()) {
+		m_scroll_remainder.setX(0.0f);
+		m_scroll_last_direction.setX(direction_x);
+	}
+	int direction_y = normalize(scroll_delta_f.y());
+	if (direction_y != m_scroll_last_direction.y()) {
+		m_scroll_remainder.setY(0.0f);
+		m_scroll_last_direction.setY(direction_y);
+	}
+
+	// Scroll distance considering previous events
+	QPointF total_delta = scroll_delta_f + m_scroll_remainder;
+	QPoint scroll_delta = QPoint(int(total_delta.x()), int(total_delta.y()));
```

Comment:
This may result in an unnecessary copy/move. You should call the constructor directly instead:

`QPoint scroll_delta{ int(total_delta.x()), int(total_delta.y()) };`

Or better yet, maybe something like this would work:
```
const QPointF scroll_delta{ std::floor(total_delta.x()), std::floor(total_delta.y()) };

// Store remainder so we can accumulate fine scroll events
m_scroll_remainder = total_delta - scroll_delta;
```

No C-Style cast, so slightly safer. More clear the rounding is intentional. There are 2 less copies.